### DISCUSSION TITLE
Enhancing Option Selection with Groups in Controls Panel

### DISCUSSION THREAD

**Discussion by max-scopp** [react-cosmos/react-cosmos#1513] [Ideas]
In situations where a select field includes a large number of options, it's common for visual clarity to diminish, making it prone to "word skipping." This means users may inadvertently select an unintended option due to the overwhelming number of choices. This challenge becomes more pronounced when the options are predominantly similar, exacerbating the difficulty of differentiation.

Essentially, when a select field becomes visually fuzzy and users experience word skipping, it heightens the risk of unintentional selection. This issue becomes even more challenging when the available options share similarities.

As a result, the user experience and efficiency of utilizing Cosmos may be compromised. The cognitive burden of navigating through a visually fuzzy select field and the increased likelihood of unintentional selections can impede the smooth workflow within the Cosmos environment. This reduction in efficiency could hinder productivity and potentially diminish the overall usability of the tool.

**Use case:**

In certain scenarios, there is a need for us to provide users with a predefined set of CSS Variables for specific components. For example, let's consider the component `<Flex gap="cssPropertyValue" />`, where users can specify the `gap` using CSS Variables like `<Flex gap="var(--vendor-spacing-big)" />`. In this case, it would be advantageous to have an improved approach for selecting CSS Variables directly from the controls panel.

Currently, the list of CSS Variables includes all options, regardless of whether they are compatible with the specific component or not. To address this use case, a potential solution is to introduce a grouping mechanism that allows users to easily differentiate between different types or categories of CSS Variables. This grouping feature would enhance the usability of the controls panel, making it simpler for users to locate and select the relevant CSS Variables that are compatible with the component they are working with.

![Screenshot 2023-05-29 at 11 32 01](https://github.com/react-cosmos/react-cosmos/assets/11851540/d4837f3d-154d-4a67-8f38-55589170b94c)

**Comment by ovidiuch** [comment]
I think this would be a useful addition.

I'll leave some code references related to which parts of the source code would require changes to implement this.

- The underlying type of the control. This is part of what we call "fixture state" — a serializable representation of the fixture state that can be manipulated remotely.

https://github.com/react-cosmos/react-cosmos/blob/087a590d5e9d6803b2592c13dd364c19c146efd0/packages/react-cosmos-core/src/fixtureState/types.ts#L67-L72

- The part of the `useSelect` hook that creates the control item in the fixture state.

https://github.com/react-cosmos/react-cosmos/blob/087a590d5e9d6803b2592c13dd364c19c146efd0/packages/react-cosmos-renderer/src/fixture/useSelect/useCreateFixtureState.ts#L29-L34

- The input component that manipulates this fixture state in the Cosmos UI.

https://github.com/react-cosmos/react-cosmos/blob/087a590d5e9d6803b2592c13dd364c19c146efd0/packages/react-cosmos-ui/src/plugins/ControlSelect/SelectValueInput.tsx#L26-L39

- The low level Select component reused throughout the Cosmos UI. You can see it only supports a basic list of string options and it would need to be extended to support [`<optgroup>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/optgroup)s.

https://github.com/react-cosmos/react-cosmos/blob/087a590d5e9d6803b2592c13dd364c19c146efd0/packages/react-cosmos-ui/src/components/inputs/Select.tsx#L70-L76

If someone wants to take charge and implement this feel free and let me know if you need more help. It's not a simple contribution by any means because you're touching 3 packages in the process. But will some luck, if you don't hit any roadblocks, it's a _doable_ outside contribution of considerable impact.

I wouldn't worry about tests for the UI changes (although changes to the Select component should be sensible because it is a reusable primitive). But one or two brief useSelect tests similar to [these](https://github.com/react-cosmos/react-cosmos/blob/main/packages/react-cosmos-renderer/src/__tests__/selects.tsx) that feature fixture state with group options would be useful.

**Comment by max-scopp** [comment]
I have completed an initial implementation, which can be found in the following GitHub pull request: [link](https://github.com/react-cosmos/react-cosmos/pull/1516). I would appreciate your feedback on any potential improvements or changes you would suggest.

Tomorrow I will have to work again, but I will try to find time this week to enhance the test coverage.
